### DISCUSSION TITLE
WANTED: Feedback on The Cookbook

### DISCUSSION THREAD

**Discussion by erondpowell** [directus/directus#16597] [Archive]
Greetings from Docs Land!  
I'm Eron, the technical writer here at Directus.

**And I need some feedback on the new [Cookbook](https://docs.directus.io/cookbook/data-models/personal-portfolio-using-m2a.html) 👨‍🍳** 

As the platform continues to scale, so have the questions! 
About half of the requests fall under the _"how-to do xyz"_ category... For example: 
- How to create a flows for-loop
- How to setup an e-commerce data model
- _etc. etc. etc.!_

In the [Cookbook](https://docs.directus.io/cookbook/data-models/personal-portfolio-using-m2a.html), each **recipe** answers a *"how-to"* question like this. Before we get carried away making 100s or 1000s of recipes, we figure we ought to release an MVP with a *handful* of recipes and ask our brilliant dev community what their thoughts are.

**Key Questions**
- Are you likely to use recipes like this for your projects? Why or why not?
- We'd _**love**_ the cookbook to be a community-driven effort. Would you be willing to contribute recipes? Why or why not?

**Bonus**
- Do you have any recipe requests?  🥘 🍗 
- Do you have any cool recipes? 👨‍🍳  👩‍🍳 

*Additional thoughts, feelings, advice, ideas, feedback and tangents on the recipes welcome* 🙏

Cheers 🥂  and thanks 🙏 !

**Comment by andy1li** [comment]
Hi Eron,

Great cookbook!

A potential bug report: idk whether it's my network connection problem or not, but I can't play the video in [Flows For Loops](https://docs.directus.io/cookbook/flows/flows-for-loop.html).

**Reply by br41nslug** [reply]
Thanks for letting us know 😄 the video was a placeholder and left there, we'll hide it until there is an actual video to show

**Comment by jdamboeck** [comment]
Sure i'd like some yumie recipes. Because some things i can not cook for myself. Like for example an recipe for deploying with docker and updating the schema when upping the container would be awesome.

As for the questions.
- Q: Are you likely to use recipes like this for your projects? 
  A :If the recipe is for something i can't figure out myself quickly.
- Q: Would you be willing to contribute recipes?
  A: Sure will if my knowledge of directus grows a little more and i know something which other users may want to cook.

Also, the effort is very much appreciated.

**Reply by erondpowell** [reply]
Thanks for the feedback!

**Comment by MentalGear** [comment]
Quick note It could use some real example repos.

**Reply by erondpowell** [reply]
> some real example repos.

Hey ya @MentalGear ! I'm not 100% sure what you're referring to. Can you elaborate?

**Comment by kevincam3** [comment]
Hi Eron,

Nice to virtually meet you. I think the Cookbook is a great idea. I didn't notice it before but I'll be going through all the items there now. I would definitely make use of a Cookbook and contribute as well because that not only helps others but helps those that contribute as well. I'm trully loving Directus. I'm going to make this the primary go to headless CMS. I think the docs are great I know it must be hard to keep them up to date. As I read through them I can see there is information missing that would be helpful and some that has changed and the docs don't reflect the new way. 

I've decided I want to help with this project. It will be the first time I'm going to be trying to help with an opensource project so I hope I can contribute with my knowledge in some way.

I would love to see receipies for making custom extensions. I'm currently working on creating a custom interface extension that will let me add the "dated_created" field that you get when you create a new collection but is not available when you connect your own database to Directus. I do know that it's easy to just duplicate that field but I'm doing this so that I can understand how doing this works so I have that power to make what every interface I want.

I would love to see this done as a recipe and also something for the ther extension types as well. Some I don't even know when I would use them like "hook" so having something like this in the Cookbook would also help when use cases for some of the features Directus has. 

Great job and hats off the the Directus team. Out of my decades of messing around with web development I don't think I've ever seem a platform as powerful and well built as Directus. I'm loving it.

**Reply by erondpowell** [reply]
> Nice to virtually meet you. 

You too Kevin! Very happy to hear you're interested in helping with Directus. If you are interested, I'd love to book an interview to deep-dive into your experience with Directus, learn a bit more about your background, and chat a bit more about Docs Land!

> I would love to see recipes for making custom extensions. 

Yeah! We plan to have some extensions guide! A doc on how to build Extension(s) is fundamentally broad in scope. 
The recipes are currently written to fit hyper-specific use-cases. So that type of extensions guide *might* go in another area, tbd. But that's def in the works.

> I've ever seem a platform as powerful and well built as Directus.

Love that!  CC @rijkvanzanten @MichelleEMadd @phazonoverload

**Reply by kevincam3** [reply]
Hi Eron,

Sure we can schedule a interview to see how I might be able to contribute/help. 
https://meetings.hubspot.com/kevincam/book

**Reply by erondpowell** [reply]
> Sure we can schedule a interview to see how I might be able to contribute/help. https://meetings.hubspot.com/kevincam/book

Okay, I went ahead and booked it and left some notes... not 100% what timezone you're in.
Feel free email me eron@directus.io or ping me in the Directus channel to followup/confirm.

**Comment by josephdpurcell** [comment]
**Are you likely to use recipes like this for your projects? Why or why not?**

Yes. Even a few thorough recipes can be enough to extrapolate to many more use cases, so even though the Cookbook may not give me the recipe I want, I can likely extrapolate from it.

Recipes offer a way to show how various features can be composed in unique ways to build even more expressive behavior, Flows and Webhooks are good examples of that since by themselves they are meaningless -- its when you compose them with other systems or features within Directus that the power of those features becomes mind boggling and exciting.

**We'd love the cookbook to be a community-driven effort. Would you be willing to contribute recipes? Why or why not?**

Yes. I enjoy writing documentation, contributing to open source, and I think Directus is a great project.

**Do you have any recipe requests?**

Yes. Detailed Flows examples! See https://github.com/directus/directus/issues/16674#issuecomment-1331210812.

Here is a list of things I would like to see as part of recipes:

* How to get an item's content into the data chain and reference it
* How to submit a notification to the current user, eg Notification op or Email op
* How to safely prepare an HTTP request body (I realized that just using the Webhook/Request URL was double encoding JSON, at least I think it was?)
* How to give user feedback when a user presses a manual trigger and it errors (I used email ops because notification op has the null/null bug, but maybe there is a synchronous way to give that feedback in the "Flow Ran Successfully" message that pops up)
* How to use a Condition op on an item's content (I never figured this out)

 **Do you have any cool recipes?**

I think what I wrote https://github.com/directus/directus/issues/16674#issuecomment-1331210812 is pretty neat: the idea is you can author content in Directus, push it out to Listmonk as a Campaign, then trigger an email campaign. The opportunities go far beyond that once you know this basic recipe that don't even include Listmonk (e.g. automatically tweeting a link to a blog post you write, sending SMS notification to a peer to review and publish an article for you).

Edit 9/30 715a ET: Itemized recipe aspects.

**Reply by erondpowell** [reply]
Thanks for the feedback! and I just read over comment #16674  we well.

Some of the stuff you've mentioned belongs in the actual docs! Some could be a recipe! Flows is also quite tricky to document every single implementation/use-case because its basically *a system to get data from anywhere, do whatever you want with it, and send it anywhere it needs to go.* 

However, we definitely plan 1-2 more big PRs to the flows docs, including:
- filling in more loose ends, to alleviate confusion similar to what you've described
- adding videos to explicitly, visually, demonstrate how each trigger and operation works

We currently have several big projects in queue in Docs Land, but we aim to close those flows PRs out within the next few months.

If you're interested, I'd love to do an interview where we talk through your project, experience with flows, pull up the docs to read through together, and deep-dive into these documentation pain-points you're experiencing. Feel free to email me at eron@directus.io or ping me on our [Discord](https://discord.gg/directus) channel `eron_powell` 

CC @phazonoverload _(He's our new Dev Rel, an utter UX wiz, and may wanna join the call too!)_

**Reply by josephdpurcell** [reply]
Yep, I'd be happy to. I DM'd you on Discord.

**Comment by raphourbe** [comment]
**Are you likely to use recipes like this for your projects? Why?**  
Yes definitely! I have been using nextjs and their repo full of examples is really nice. I use it quite often. So I would definitely check the cookbook first before starting something new.

**We'd love the cookbook to be a community-driven effort. Would you be willing to contribute recipes? Why or why not?**
Yes. I have actually written some how-to for directus on my blog. It's kind of my own documentation for the SaaS I created using Directus and Stripe Billing for my company. Please take what you need!

**Do you have any cool recipes?** 👨‍🍳 👩‍🍳
You tell me :) Here are my current posts [https://rphl.dev/](https://rphl.dev/)

- Building a newsletter subscription with Directus.
- Listen to Stripe webhooks with Directus.
- Integrate Stripe to Directus' flows.
- Binding Directus auth and Next.js middleware to protect your app.
- Using Directus js-sdk in a Next.js app with several useSWR calls.
- Using Directus with Next-Auth to handle authentication in your frontend.
- Building a SaaS with Directus and Stripe (4-part series)

**Reply by kevincam3** [reply]
Wow, this is some good stuff. We definitely need to get this stuff over into the Cookbook so we can start building a single place to find all these goodies.

**Reply by saqibvohra93** [reply]
@raphourbe, your blog has got superb, amazingly-well written content. Thank you. Really helping a lot with different functionalities around Directus. 

It would be an added help if we can get some blogs around extensions as well in your leisure time.

Thanks

**Reply by erondpowell** [reply]
@raphourbe I've seen your work before, these posts are _amaaaaaazing_! 
If you're still willing & have time in the future, we'll def be in touch about getting some of those refactored into the cookbook.

**Reply by raphourbe** [reply]
Thank you all 😄 

@erondpowell Sure, let me know.

@saqibvohra93 You have the post "Listen to Stripe webhooks with Directus." which shows a quick implementation of two kinds of extensions: `hook` & `endpoint`. I will soon cover an in-depth integration of Stripe webhook with an `endpoint` extension (it's the last part of my Saas with Directus/Stripe series). And if it is of interest, I can also do a little post on how I connected an existing API I have to Directus using an `endpoint`.

**Reply by saqibvohra93** [reply]
Absolutely. The Stripe series helped greatly with understanding about some of the API extensions, and a few other areas. Looking forward to the sequel.

Also, I have a very vague understanding of App extensions so far. If feasible, that would be a great help for me, and I think for the Directus community as well. Thanks

**Reply by raphourbe** [reply]
I have never used Vue.js so it might take time before I am able to show an example.

I have a new post about creating a custom operation though: https://rphl.dev/blog/enhancing-directus-send-mail-operation (it is based on this post, where you also find my answer https://github.com/directus/directus/discussions/16726).

It shows how to develop an enhanced Send Mail Operation, where you have Cc, Bcc and template selection added. There is an `app.js` file that renders the Operation in the Flow part of the app.

However, to my comprehension, after reading the doc the modification of the App is something way different.

**Reply by saqibvohra93** [reply]
Thank you, huge respect for you sir! I was actually looking into sending dynamic emails using a 3rd party application. But with the help of this article, I think I might be able to do it on the platform, saving me some $$. 🙌🙌

**Reply by LasseRosenow** [reply]
I also think the posts of raphourbe are really amazing and so detailed. 

If the cookbook has the same structure as this, it would truly be amazing :) Oh please integrate them, if it is allowed :)

**Reply by raphourbe** [reply]
Building a SaaS with Directus and Stripe (4-part series) is now complete 😄 The last post is quite long, be prepared...

It's available [here](https://rphl.dev/blog/building-saas-with-directus-and-stripe-part4). I explain in it the implementation of a Hook extension and an Endpoint extension to retrieve Stripe's webhooks.

**Reply by erondpowell** [reply]
CC @phazonoverload who I believe is also looking to make some tutorials around extensions fairly soon.

**Comment by saqibvohra93** [comment]
Are there any plans to include recipes related to extensions in the Cookbook?

I am looking for a guide to creating a custom module with configurable navigation options.

**Reply by erondpowell** [reply]
That's a really good question 

Currently, the _categories_ a recipe could fall in are as follows.
- projects (e.g. config directus for ecommerce. these would be composed of multiple "bite-size" recipes.)
- integrations (e.g. link Directus to Nuxt.js frontend)
- data model
- users roles permissions
- flows
- websockets (when released)
- api/sdk

(Keep in mind the cookbook format is still experimental... It's in *"beta testing"* 😆  and up for changes over the next few months)

( CC @rijkvanzanten @phazonoverload )
Every category in the list, _except integrations which involve outside services/frontends/etc_, includes tasks that happen on or in the database.

Extensions happen on the core platform... I personally haven't worked with Extensions yet, but I think we might need another content format for extensions guides. To go a step deeper:

- We definitely need to improve the existing extensions documentation.
- As far as extensions "guides": well, recipes are hyper-specific. For an extension, once you create a hyper-specific guide/recipe, I imagine you'd release your extension to the public. So I reckon we'd want a more *open-ended or conceptual* content format for Extensions "guides", so that we can *teach you to fish, instead of feeding you fish*. But idk, I haven't spent any actual time scoping that yet.

**Reply by saqibvohra93** [reply]
Excited about the cookbook. However, I have been able to grasp concepts related to most of the categories(excluding some parts of Flow like Scripts), with the help of currently available documentation and some quick play-around.

The challenges I am facing are around the extensions. Due to a lack of detailed documentation(just my opinion as someone with basic experience in core web dev) around it, I am unable to grasp the ideas and implementations related to extensions.

Cookbook/enhanced guides around extensions will allow me to build experimental knowledge related to extensions. This will open the possibilities for integrating other open-source platforms(e.g: Metabase), while also allowing me to solution for use cases better, by understanding the possibilities and limitations.

I see and am setting up Directus on my current project as something much more than the data collaboration app. It's just missing a very few features that I need, which I am confident, will be able to resolve with the help of extensions, opening the door to solving challenges in a lot of key areas, the enterprises are facing today.

Thus, a more detailed guide and multiple recipes on each extension will be greatly helpful and will increase community contribution to the Directus platform as well.

Hope this helps in planning the docs.

P.S.: Not a very good writer, thought of sharing my experience with the Directus using the material available. Would love to chat if helpful in anyway.

**Reply by erondpowell** [reply]
> Due to a lack of detailed documentation

You're absolutely correct. it is lacking. We plan to begin making serious updates *very soon*.
It just may or may not be *in the cookbook*.

> P.S.: Not a very good writer 

In my personal opinion as a full-time writer, good writing is not about fancy words and language-- but about about making your point clear and with strong concrete examples to back it up. and to that you excel

**Reply by erondpowell** [reply]
oh, I almost forgot!

> Would love to chat if helpful in anyway.

When we go to begin updating the extensions, I'll try to loop back and get an interview with you tog et your experience. If you still have time at that point in the future.

**Reply by saqibvohra93** [reply]
Thanks @erondpowell. It would be an absolute pleasure any day.

**Comment by SonNg2k** [comment]


**Reply by kevincam3** [reply]
It's funny because while it's greyed out if you click it, it does select it. Give it a try. It worked for me.

**Reply by SonNg2k** [reply]
It should be mentioned in the doc that in order for the Tree View setup to work, the `parent_id` field must be created at the 'Foreign Key' field in the Tree View Interface. It will not work if you create the `parent_id` field before creating the Tree View O2M alias field. There is already a [demonstration video](https://github.com/directus/directus/discussions/13376#discussioncomment-2802662) from a Directus core developer

<img width="792" alt="image" src="https://user-images.githubusercontent.com/29955043/206135879-be6a80d3-dbf5-4b33-bd04-a35d95ed71d7.png">

**Reply by erondpowell** [reply]
If you use the dropdown it is grayed out and doesn't work. 
If you manually type in the name of an existing input, it works! 
Likely a small bug in the ui. (CC @azrikahar )

I'm making a video to accompany this recipe today.
I'll be sure to provide better instructions @SonNg2k

**Reply by erondpowell** [reply]
> It's funny because while it's greyed out if you click it, it does select it. Give it a try. It worked for me.

Can you send a video of that? because I couldn't click it. But I could type it in.

**Reply by kevincam3** [reply]
Actually you might be right. I think I did type it in. I knew I got it to work.
